tests/test-awful-rules.lua

```diff
@@ -0,0 +1,84 @@
+local awful = require("awful")
+local test_client = require("_client")
+
+local callback_called = false
+local has_spawned = false
+
+-- Add a new rule --TODO add to awful.rules, convert to metatable, add clear, support enable/disable
+local function add_rule(rule)
+    table.insert(awful.rules.rules, rule)
+end
+
+-- Helper function to search clients
+local function get_client_by_class(class)
+    for _, c in ipairs(client.get()) do
+        if class == c.class then
+            return c
+        end
+    end
+end
+
+-- Test callback and floating
+add_rule {
+    rule       = { class = "rule_float" },
+    properties = { floating=true },
+    callback   = function(c) callback_called = true end
+}
+
+-- Test ontop
+add_rule {
+    rule       = { class = "rule_ontop" },
+    properties = { ontop=true },
+}
+
+-- Test placement
+add_rule {
+    rule       = { class = "rule_placement" },
+    properties = {
+        floating  = true,
+        placement = "bottom_right"
+    },
+}
+
+-- Create many clients with custom classes and names
+local function spawn_clients(count)
+    if count <= 1 and not has_spawned and #client.get() < 1 then
+        test_client("rule_float"    , "Foo"       )
+        test_client("rule_ontop"    , "Foo"       )
+        test_client("rule_placement", "Foo"       )
+        test_client("rule_bytitle"  , "TITLE1234" )
+        test_client("rule_new_tag"  , "Foo"       )
+        test_client("rule_tags"     , "Foo"       )
+    elseif #client.get() >= 6 then
+        -- Set tiled
+        awful.layout.inc(1)
+        return true
+    end
+end
+
+local function test_rules()
+    -- Test if callbacks works
+    assert(callback_called)
+
+    -- Make sure "smart" dynamic properties are applied
+    assert(get_client_by_class("rule_float").floating)
+    local geo = get_client_by_class("rule_float"):geometry()
+    --assert(geo.width == 100 and geo.height == 100) --FIXME height=100+workarea.y+border_width
```

Comment:
@psychon Is that normal? The gtk client ask for 100x100, but awesome gives 100x118 here